perf(lexer): only check for hashbang at start of file

overlookmotel opened the pull request:
Small optimization to lexer. A hashbang can only appear at very start of file, so only check for hashbang when getting first token. This streamlines the byte handler for `#`, because a `#` anywhere else can only be a private identifier.

Note: `self.token.set_is_on_new_line(true);` in `read_hashbang_comment` is not required, because it's always `true` already.

camc314 reviewed (commented):
nice 💪

overlookmotel commented:
Hmm. I'm not sure that benchmark is right. First version made no difference at all. I may have made a mistake.

Boshen commented:
### Merge activity

* **Aug 10, 7:32 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
perf(lexer): only check for hashbang at start of file (#12521)

Small optimization to lexer. A hashbang can only appear at very start of file, so only check for hashbang when getting first token. This streamlines the byte handler for `#`, because a `#` anywhere else can only be a private identifier.

Note: `self.token.set_is_on_new_line(true);` in `read_hashbang_comment` is not required, because it's always `true` already.
